### ISSUE TITLE
False positive detection likely caused by aggressive inlining in custom kernel build

### ISSUE THREAD

**Issue by sempervictus** [lkrg-org/lkrg#166]
![image](https://user-images.githubusercontent.com/1331084/154820449-29eba796-9aeb-4b14-a76f-b4abca5c947d.png)
^^ that's on the 5.10.101 build which has all of the exports marked as suggested (no errors at load-time anyway).

**Comment by solardiz** [comment]
Nasty. This illustrates another rather fundamental issue we're facing lately: even when we do find and hook the symbol, there might be an inlined copy (or several) of the same code as well, and we wouldn't hook that, resulting in inconsistent LKRG state (it was supposed to be updated by all uses of the function, but is in fact getting updated by only some or possibly even not updated at all if the function/symbol we forced the compiler to preserve with `EXPORT_SYMBOL` became dead code and not code the rest of the kernel actually calls rather than inlines).

No easy fix, sorry!

**Comment by solardiz** [comment]
In the case of your kernel build, it looks like we have this problem for at least two symbols, and I guess at least one of those wasn't among the forced `EXPORT_SYMBOL` ones (not in `add-exports.sh`).

**Comment by sempervictus** [comment]
Any idea what the missing symbol might be so i could add it to the script? Doesnt sound like it'll help the problem you describe, but at least might as well squash one bug while we're here.

**Comment by solardiz** [comment]
No, I don't think we should be adding one more symbol to the exports because of this, nor would it be expected to help.

My current best idea for a workaround is to be adding `noinline` (Linux kernel specific macro) or `__attribute__((__noinline__))` to right before every function that we hook and that may potentially be inlined, which you can do with `sed`. You can test this approach by doing it for the two seccomp-related symbols - then see if the seccomp-related false positive that killed `(ogrotate)` (why that weird name?) on your system would be gone.

This would be instead of adding the exports - we don't actually rely on the exports anyway, looking those symbols up in a hackish way instead - we just need the functions to be preserved as such and not inlined.

**Comment by Adam-pi3** [comment]
@solardiz Adding `__attribute__((__noinline__))` should help. In theory, `EXPORT_SYMBOL` can't be inline so this should work too. However, I remember that it was not always the case and compiler could generate unpredictable results:

https://lore.kernel.org/lkml/1368086323-9412-2-git-send-email-yefremov.denis@gmail.com/

**Comment by solardiz** [comment]
My understanding is that exporting a symbol takes and records its address, causing a non-inlined copy of the function to be preserved, but it does not prevent the compiler from also generating inlined copies of the function's body or its parts. Explicit `noinline` should achieve the latter.

**Comment by solardiz** [comment]
The screenshot here shows at least two issues:
1. seccomp filter pointer "corruption" for `(ogrotate)`.
2. Unexpected differences in `fsgid`, `cred`, `real_cred` for `sshd`.

For seccomp, we've since reworked our hooking and logic, so this report from 2022 is invalidated (maybe the current LKRG would have the same issue on that kernel or not, but the old testing/report is of no use now either way).

For `fsgid`, `cred`, `real_cred`, I am puzzled as to why only `fsgid` would be different. We do separately hook `__sys_setfsgid`, so if that failed, it could be the reason, but this function doesn't look significantly different from several others that we hook, so it's surprising only this one would be affected. We didn't rework our code for this so extensively as to necessarily invalidate the testing/report here, yet I also see no reasonable action we can take based on this now. I don't recall us receiving any other reports with exactly this symptom.

The screenshot also says `Arch Linux 5.10.101`. Well, I've tested recent LKRG (as in this repo) on recent Arch Linux's kernel build over the recent few weeks, and it works just fine.

So I'm closing this issue.
